sled-storage/src/manager.rs

```diff
@@ -986,12 +1291,49 @@ impl StorageManager {
         }
     }
 
-    // Ensures a dataset exists within a zpool, according to `config`.
+    // Invokes [Self::ensure_dataset] and also ensures the dataset has an
+    // expected UUID as a ZFS property.
+    async fn ensure_dataset_with_id(
+        &mut self,
+        zpool: &ZpoolName,
+        id: DatasetUuid,
+        config: &SharedDatasetConfig,
+        details: &DatasetCreationDetails,
+    ) -> Result<(), Error> {
+        self.ensure_dataset(zpool, config, details).await?;
+
+        // Ensure the dataset has a usable UUID.
+        if let Ok(id_str) = Zfs::get_oxide_value(&details.full_name, "uuid") {
+            if let Ok(found_id) = id_str.parse::<DatasetUuid>() {
+                if found_id != id {
+                    return Err(Error::UuidMismatch {
+                        name: details.full_name.clone(),
+                        old: found_id.into_untyped_uuid(),
+                        new: id.into_untyped_uuid(),
+                    });
+                }
+                return Ok(());
+            }
+        }
+        Zfs::set_oxide_value(&details.full_name, "uuid", &id.to_string())?;
```

Comment:
Hmm just a thought here, but if we create a dataset and then fail to set the uuid we currently have no way of cleaning up the half setup dataset do we? Probably not a blocker for this PR but something we might want to file an issue for so we don't lose track of it.

Reply:
Filed https://github.com/oxidecomputer/omicron/issues/7009